text/0011-utxo-model/0011-utxo-model.md

```diff
@@ -0,0 +1,194 @@
+used to give coins a *meaning*.
+
+A **color** is simply a random sequence of bytes that can be set by the user and which is retained when being
+transferred:
+
+``Color = Array<byte>``
+
+Next to *user defined* colors, there are two *builtin color* values that carry a special meaning:
+
+- ``COLOR_IOTA = Array<byte>(0, 0, 0, ..., 0)`` represents the *base color* of uncolored coins.
+
+- ``COLOR_MINT = Array<byte>(255, 255, 255, ..., 255)`` represents a color which is being replaced by the
+  ``Transaction ID`` before being booked as an output. It is consequently used to "mint" new colored coins.
+
+A **colored balance** is accordingly a combination of a numeric balance with its corresponding color:
+
+``ColoredBalance = {Balance: uint64, Color: Color}``
+
+### Outputs
+
+An Output is a container for the colored balances that were spent by a particular transaction to a certain address. In
+addition to the list of colored balances and its ID, we store an ``OPCode`` and a corresponding ``OPCodeMetadata``.
+
+``Output = {ID: OutputID, Balances: Array<ColoredBalance>, OpCode: OPCode, OPCodeMetadata: OPCodeMetadata}``
+
+### OPCodes
+
+Currently, an output is *unlocked* if there is a valid signature for the address that holds the output, and it is not
+possible to define alternative conditions.
+
+There are however use cases where being able to define different conditions might be useful. A famous example are hash
+locks and time locks, that are used to create
+[Hashed Timelock Contracts](https://en.bitcoinwiki.org/wiki/Hashed_Timelock_Contracts) enabling things like atomic swaps
+and 2nd layer scaling solutions like the Lightning Network.
+
+The opcode is a single byte that is mapped to a specific unlock method. This will allow us to switch between different
+unlock conditions which introduces some very basic smart contract capabilities on the base layer.
+
+``OPCode = byte``
+
+Since some opcodes might require certain parameters (like a timestamp or a hash), we define a metadata field, that holds
+a marshaled version of these parameters.
+
+``OPCodeMetadata = Array<byte>`` 
+
+As long as these checks are not much more *expensive* than signature checks, it does not introduce problems but
+massively extends the functionality of the protocol. We have to however be very careful with adding new conditions, to
+not accidentally introduce potential bottlenecks.
+
+In contrast to bitcoin, we do not add a fully scriptable outputs (because that could result in long execution times),
+but we offer predefined methods for different use cases.
+
+For now, we simply reserve the additional fields in the output but hard code the OPCode to be ``SIG_CHECK = OPCode(0)``.
+
+### Conflict Detection
+
+The biggest benefits of this ledger state is the fast and easy detection of double spends and invalid transaction:
+
+- A transaction is valid, if the sum of the moved funds is ``0`` and it only spends colors in the amounts that were
```

Comment:
There is more to validity:
- There must be at least one output (and at least one input)
- the output referenced by each OutputID in the input list must be valid and exists.
- the total output sum (per color) must be equal to the sum (per color) over all outputs
- only one output per address
- What about 0-value inputs/outputs are they valid?
- ...